sunpy/physics/differential_rotation.py

```diff
@@ -84,138 +89,315 @@ def diff_rot(duration: u.s, latitude: u.deg, rot_type='howard', frame_time='side
-    >>> solar_rotate_coordinate(c, end_time)
+    >>> from sunpy.time import parse_time
+    >>> start_time = parse_time('2010-09-10 12:34:56')
+    >>> end_time = parse_time('2010-09-10 13:34:56')
+    >>> c = SkyCoord(-570*u.arcsec, 120*u.arcsec, obstime=start_time, frame=Helioprojective)
+    >>> solar_rotate_coordinate(c, time=end_time)
     <SkyCoord (Helioprojective: obstime=2010-09-10T13:34:56.000, rsun=695700.0 km, observer=<HeliographicStonyhurst Coordinate for 'earth'>): (Tx, Ty, distance) in (arcsec, arcsec, km)
-        (-562.37689548, 119.26840368, 1.50083152e+08)>
-
+    (-562.89557218, 119.31528517, 1.50084836e+08)>
+    >>> new_observer = get_body("earth", end_time)
+    >>> solar_rotate_coordinate(c, observer=new_observer)
+    <SkyCoord (Helioprojective: obstime=2010-09-10T13:34:56.000, rsun=695700.0 km, observer=<HeliographicStonyhurst Coordinate for 'earth'>): (Tx, Ty, distance) in (arcsec, arcsec, km)
+        (-562.89557218, 119.31528517, 1.50084836e+08)>
     """
+    # Check the input and create the new observer
+    new_observer = _get_new_observer(coordinate.obstime, observer, time)
+
+    # The keyword "frame_time" must be explicitly set to "sidereal"
+    # when using this function.
+    diff_rot_kwargs.update({"frame_time": "sidereal"})
 
     # Calculate the interval between the start and end time
-    interval = (
-        parse_time(new_observer_time) - parse_time(coordinate.obstime)).to(u.s)
+    interval = (new_observer.obstime - coordinate.obstime).to(u.s)
 
     # Compute Stonyhurst Heliographic co-ordinates - returns (longitude,
     # latitude). Points off the limb are returned as nan.
-    heliographic_coordinate = coordinate.transform_to('heliographic_stonyhurst')
+    heliographic_coordinate = coordinate.transform_to(HeliographicStonyhurst)
 
     # Compute the differential rotation
     drot = diff_rot(interval, heliographic_coordinate.lat.to(u.degree), **diff_rot_kwargs)
 
-    # Rotate the input co-ordinate and update the observer
-    heliographic_rotated = SkyCoord(
-        heliographic_coordinate.lon + drot,
-        heliographic_coordinate.lat,
-        obstime=new_observer_time,
-        observer=new_observer_location,
-        frame=frames.HeliographicStonyhurst)
+    # Rotate the input co-ordinate as seen by the original observer
+    heliographic_rotated = SkyCoord(heliographic_coordinate.lon + drot, heliographic_coordinate.lat,
+                                    obstime=coordinate.obstime, observer=coordinate.observer,
+                                    frame=HeliographicStonyhurst)
 
-    # Return the rotated coordinates to the input coordinate frame
-    return heliographic_rotated.transform_to(coordinate.frame.name)
+    # Calculate where the rotated co-ordinate appears as seen by new observer,
+    # and then transform it into the co-ordinate system of the input
+    # co-ordinate.
+    return heliographic_rotated.transform_to(new_observer).transform_to(coordinate.frame.name)
```

Comment:
I don't *think* this is quite what we need here. Do we not need the output frame to have the `observer` set to the `new_observer`?

This is a bit tricky if you want essentially the same frame as `coordinate` but with a new observer?

Reply:
I don't think I quite understand.  `heliographic_rotated` adds on the amount of rotation as seen from `coordinate.observer`, and then that point on the Sun as seen from there is transformed to another point because of the change of observer.  Doesn't the first `transform_to` change the observer property?

Reply:
Yeah but what's the second transform trying to do? You are using all the defaults for the last frame?

Reply:
Oh I see this is all SkyCoord, so this might actually do what we need, still feels weird though.

Reply:
The second one is just to re-calculate the coordinate back in to the type it was passed in as.